lib/src/search.dart

```diff
@@ -0,0 +1,203 @@
+    var jsonIndex = (jsonDecode(text) as List).cast<Map<String, Object>>();
+    var indexList = <IndexItem>[];
+    var packageOrder = <String>[];
+    for (var entry in jsonIndex) {
+      if (entry.containsKey(packageOrderKey)) {
+        packageOrder.addAll(entry[packageOrderKey] as List<String>);
+      } else {
+        indexList.add(IndexItem.fromMap(entry));
+      }
+    }
+    return Index(packageOrder, indexList);
+  }
+
+  int packageOrderPosition(String packageName) {
+    if (packageOrder.isEmpty) return 0;
+    var index = packageOrder.indexOf(packageName);
+    return index == -1 ? packageOrder.length : index;
+  }
+
+  List<IndexItem> find(String rawQuery) {
+    if (rawQuery.isEmpty) {
+      return [];
+    }
+
+    var query = rawQuery.toLowerCase();
+    var allMatches = <({IndexItem item, _MatchPosition matchPosition})>[];
+
+    for (var item in index) {
+      void score(_MatchPosition matchPosition) {
+        allMatches.add((item: item, matchPosition: matchPosition));
+      }
+
+      var lowerName = item.name.toLowerCase();
+      var lowerQualifiedName = item.qualifiedName.toLowerCase();
+
+      if (lowerName == query ||
+          lowerQualifiedName == query ||
+          lowerName == 'dart:$query') {
+        score(_MatchPosition.isExactly);
+      } else if (query.length > 1) {
+        if (lowerName.startsWith(query) ||
+            lowerQualifiedName.startsWith(query)) {
+          score(_MatchPosition.startsWith);
+        } else if (lowerName.contains(query) ||
+            lowerQualifiedName.contains(query)) {
+          score(_MatchPosition.contains);
+        }
+      }
+    }
+
+    allMatches.sort((a, b) {
+      // Exact match vs substring is king. If the user has typed the whole term
+      // they are searching for, but it isn't at the top, they cannot type any
+      // more to try and find it.
+      var comparison = a.matchPosition - b.matchPosition;
+      if (comparison != 0) {
+        return comparison;
+      }
+
+      // Prefer packages higher in the package order.
+      comparison = packageOrderPosition(a.item.packageName) -
+          packageOrderPosition(b.item.packageName);
+      if (comparison != 0) {
+        return comparison;
+      }
+
+      // Prefer top-level elements to library members to class (etc.) members.
+      comparison = a.item._scope - b.item._scope;
+      if (comparison != 0) {
+        return comparison;
+      }
+
+      // Prefer non-overrides to overrides.
+      comparison = a.item.overriddenDepth - b.item.overriddenDepth;
+      if (comparison != 0) {
+        return comparison;
+      }
+
+      // Prefer shorter names to longer ones.
+      return a.item.name.length - b.item.name.length;
+    });
+
+    return allMatches.map((match) => match.item).toList();
+  }
+}
+
+class IndexItem {
+  final String name;
+  final String qualifiedName;
+
+  // TODO(srawlins): Store the integer of in the package order instead of this
```

Comment:
nit: `the index of the package in package order` or similar

Reply:
Thanks! Done